### PR TITLE
add searchable songs table & necessary query

### PR DESCRIPTION
**Pull request by leethomas**
EDIT: Merge this first: https://github.com/GracepointMinistries/GraceTunes/pull/5

Songs index looks like this now:

Looks ugly, but it's very customizable. @josephli036 said he'd make it drop dead gorgeous. I'll help as well.

The search bar updates the table as you type.

### CODE REVIEW

File: app/controllers/songs_controller.rb [hunk 1]
```diff
@@ -1,5 +1,34 @@
 class SongsController < ApplicationController
   def index
+    respond_to do |format|
+      format.json do
+        songs = Song.all
+        total_songs = songs.count
+
+        if params[:sSearch].present?
```
Comment (patrickf3139) [inline app/controllers/songs_controller.rb]: I'm guessing jQuery or the gem you are using uses camel case for variable names? Just wondering why you didn't use `s_search` instead.

Comment (leethomas) [inline app/controllers/songs_controller.rb]: Yeah that's a parameter that's sent by the gem so we can't change it.

File: app/controllers/songs_controller.rb [hunk 2]
```diff
@@ -1,5 +1,34 @@
...
+          search = {
+            query: "%#{params[:sSearch]}%",
+            rQuery: Regexp.escape(params[:sSearch])
+          }
+
+          songs = songs.where('name LIKE :query OR artist LIKE :query OR
```
Comment (patrickf3139) [inline app/controllers/songs_controller.rb]: Having `songs = Song.all` and then `songs.where()` here seems redundant. You can probably combine it into `songs = Song.where()` here can't you I think. Also, it seems this query logic should be abstracted away for isolation and DRYness. I'll work on that.

Comment (leethomas) [inline app/controllers/songs_controller.rb]: The idea is that since it's the index page you want to load all of the songs (paginated) so that the user can browse through them. If they choose to filter then we narrow down the query. That's why the `songs.where` is nested in the block which checks if the search is present. 

As for abstracting the query logic, sounds good to me. I don't know how you want it done so I'll let you handle that.

Comment (patrickf3139) [inline app/controllers/songs_controller.rb]: Ah I missed the if statement. I see.

File: app/controllers/songs_controller.rb [hunk 3]
```diff
@@ -1,5 +1,34 @@
 class SongsController < ApplicationController
   def index
+    respond_to do |format|
```
Comment (patrickf3139) [inline app/controllers/songs_controller.rb]: I'm trying to understand the design and usage of the jQeury database gem. Correct me if I'm wrong: on the initial page load, Rails will not send any song data to the browser. But the JS on the page will get triggered and send a GET request and hit this code, which will then return all the songs? So the song data is never directly sent with the HTML body but retrieved when the gem requests it?

Comment (leethomas) [inline app/controllers/songs_controller.rb]: Exactly. Song data is grabbed asynchronously after the page loads.

File: app/finders/songs_finder.rb [hunk 4]
```diff
@@ -1,26 +1,26 @@
 class SongsFinder
-  
+
   def self.search(keywords)
     return [] if keywords.blank?
-    weighted_query = ActiveRecord::Base::sanitize(">(#{keywords})")
-    unweighted_query = ActiveRecord::Base::sanitize(keywords)
+    weighted_query = ActiveRecord::Base::sanitize(">(#{keywords}*)")
```
Comment (patrickf3139) [inline app/finders/songs_finder.rb]: Excellent, I was thinking that only the last term should be partial match. I think you'll also have to strip the space at the end, lest you end up with something like this: `term1 term2 *`.

**Comment by patrickf3139** [comment]
Also, please add some tests!

File: test/controllers/songs_controller_test.rb [hunk 5]
```diff
@@ -1,19 +1,19 @@
...
-  # end
-
-  # test "should get edit" do
-  #   get :edit
-  #   assert_response :success
-  # end
+  test "index should retrieve relevant search results" do
+    get :index, sSearch: 'hand', format: :json, xhr: true
+    songs = JSON.parse(@response.body)['data']
+    assert_equal(songs.length, 2)
```
Comment (patrickf3139) [inline test/controllers/songs_controller_test.rb]: This test is too flaky because as we add new song fixtures to songs.yml, there might eventually be another one with "hand" in the lyrics and suddenly this search breaks. Instead, I would do two `assert_contains` to check that the two songs currently in there are in there next time.

File: test/finders/songs_finder_test.rb [hunk 6]
```diff
@@ -34,6 +35,11 @@ class SongsFinderTest < ActionController::TestCase
...
 
   test 'retrieve all returns many songs' do
-    assert(SongsFinder.retrieve_all.count > 7) 
+    assert(SongsFinder.retrieve_all.count > 7)
+  end
+
+  test 'should partially match' do
+    assert_includes(partial_word_results, songs(:hands_to_the_heaven))
+    assert_includes(partial_word_results, songs(:glorious_day))
   end
```
Comment (patrickf3139) [inline test/finders/songs_finder_test.rb]: How about one more test to check that the partial matching only works on the last term?

File: test/controllers/songs_controller_test.rb (hunk 5 above)
Comment (patrickf3139) [inline test/controllers/songs_controller_test.rb]: Also please write one more test for index with empty params (simulating initial page load).

File: app/controllers/songs_controller.rb [hunk 7]
```diff
@@ -1,5 +1,25 @@
...
+      format.json do
+        songs = Song.all
+        total_songs = songs.count
+        songs = SongsFinder.search(params[:sSearch]) if params[:sSearch].present?
+        
+        song_data = {
+          draw: params[:draw].to_i,
+          recordsTotal: total_songs,
+          recordsFiltered: total_songs - songs.count,
```
Comment (patrickf3139) [inline app/controllers/songs_controller.rb]: Just curious, what is this recordsFiltered field for?

File: test/finders/songs_finder_test.rb (hunk 6 above)
Comment (leethomas) [inline test/finders/songs_finder_test.rb]: Not sure what you mean. This test is already doing that. If you're wondering why records still show after you've typed a relevant search + gibberish at the end, it's because you've required that at LEAST one word in the phrase matches (OR condition).

So even if you type gibberish at the end of a relevant search, as long as the previous words matched results will still show.

File: test/controllers/songs_controller_test.rb (hunk 5 above)
Comment (leethomas) [inline test/controllers/songs_controller_test.rb]: Already have that:
```rb
  test "should get index" do
    get :index
    assert_response :success
  end
```

File: app/controllers/songs_controller.rb (hunk 7 above)
Comment (leethomas) [inline app/controllers/songs_controller.rb]: data tables param: https://datatables.net/manual/server-side

```
Total records, after filtering (i.e. the total number of records after filtering has been applied - not just the number of records being returned for this page of data).
```

File: test/finders/songs_finder_test.rb (hunk 6 above)
Comment (patrickf3139) [inline test/finders/songs_finder_test.rb]: Ah let me explain: make sure the search is effectively `term1 term2 partial-term3*` and not 'term1* term2* term3*'

File: test/controllers/songs_controller_test.rb [hunk 8]
```diff
@@ -1,19 +1,22 @@
-require 'test_helper'
+require "test_helper"
```
Comment (patrickf3139) [inline test/controllers/songs_controller_test.rb]: Just curious: is there a ruby standard or convention that says to use " over ' in these cases?

File: test/controllers/songs_controller_test.rb [hunk 9]
```diff
@@ -1,19 +1,22 @@
...
 
-  # test "should get new" do
-  #   get :new
-  #   assert_response :success
-  # end
+  test "index should retrieve relevant search results" do
+    get :index, sSearch: "hand", format: :json, xhr: true
+    songs = JSON.parse(@response.body)["data"]
+    song_names = songs.map { |s| s["name"] }
+    assert_includes(song_names, songs(:hands_to_the_heaven)[:name])
```
Comment (patrickf3139) [inline test/controllers/songs_controller_test.rb]: You can just do `assert_includes(songs, songs(:hands_to_the_heaven))` without going the roundabout way of comparing names, which is flaky anyway as multiple songs can have the same name.

**Comment by patrickf3139** [comment]
I'm proud of you for more good work! After you add the partial words test and fix the roundabout test, feel free to squash and merge! :)

File: test/finders/songs_finder_test.rb [hunk 10]
```diff
@@ -3,7 +3,9 @@
 class SongsFinderTest < ActionController::TestCase
   single_word_results = SongsFinder.search "relevant"
   multi_word_results = SongsFinder.search "truth live life hands"
-  
+  partial_word_results = SongsFinder.search 'hand'
+  all_partial_word_results_2 = SongsFinder.search 'go b prais io'
```
Comment (patrickf3139) [inline test/finders/songs_finder_test.rb]: this looks like a one off variable/list, so I'd just use it in the test down below

File: test/controllers/songs_controller_test.rb (hunk 8 above)
Comment (leethomas) [inline test/controllers/songs_controller_test.rb]: At work and every other circle I've used single quotes, but I don't think there's a convention. I switched it to double quotes because the rest of the file was already using them.

**Comment by patrickf3139** [comment]
Sweet merged! Good job Lee. And congrats on your first completed PR for GTS!
